### PR TITLE
[dynamo][guards-cpp-refactor] PythonLambdaGuardAccessor

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120726
* #120739
* #120673
* __->__ #120730
* #120864

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -2440,6 +2440,54 @@ class GlobalWeakRefGuardAccessor : public GuardAccessor {
...
+    return result;
+  }
+
+  GuardDebugInfo check_verbose_nopybind(
+      PyObject* obj) override { // borrowed ref
+    PyObject* x = PyObject_CallOneArg(_accessor_fn.ptr(), obj); // new ref
+    if (x == nullptr) {
+      // The accessor function failed.
+      PyErr_Clear();
+      return GuardDebugInfo(false, std::string("Call failed ") + repr(), 0);
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:2476]: We should put the exception message in this error.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/guards.cpp (hunk 1 above)
Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp:2476]: Done

File: torch/csrc/dynamo/guards.cpp [hunk 2]
```diff
@@ -842,6 +856,22 @@ class LAMBDA_GUARD : public LeafGuard {
...
+      return GuardDebugInfo(false, exc_message, 0);
+    }
+    bool result = PyObject_IsTrue(x);
+    Py_DECREF(x);
+    if (result) {
+      return GuardDebugInfo(true, 0);
+    }
+    return GuardDebugInfo(false, verbose_code_parts(), 0);
+  }
+
```
Comment (anijain2305) [inline torch/csrc/dynamo/guards.cpp:874]: Added one for LAMBDA_GUARD as well.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**